### DISCUSSION TITLE
Best settings for torrents

### DISCUSSION THREAD

**Discussion by jiripospisil** [juicedata/juicefs#2328] [Q&A] [answered]
Hello, I have a low-end machine seeding a few torrents stored in an S3 compatible storage (scw) using S3QL. I'm trying to move to JuiceFS but so far I'm getting unusable download speeds compared to S3QL (~12 MiB -> ~20 KiB). There has to be some overhead because the storage doesn't support random writes but not like this. I'm either hitting a bug or there's some issue with my setup. I have experimented with pretty much every option I could find but no luck.

```
juicefs mount -o allow_other,writeback_cache --no-usage-report --cache-size 1500 --writeback "sqlite3://<..>.db" /mnt/asdf
```

The block size is 4 MiB. There also seems to be a lot of activity for a 200 MiB torrent.

<img width="548" alt="image" src="https://user-images.githubusercontent.com/20820/177015904-fd9ec1f4-e490-4f2d-8a6b-8adf0dbe9cd5.png">

When I'm writing sequentially (e.g. `head -c 500M /dev/zero | pv > /mnt/asdf/test`), I'm getting speeds around 30 MiB/s). Any ideas? Thanks.

v1.0.0-rc2
Arch Linux 5.18.8

**Comment by xiaogaozi** [comment, marked as answer]
Hi, have you tried use [`juicefs profile`](https://juicefs.com/docs/community/operations_profiling) command to analyze the performance problem? And it seems like the latency of metadata engine is too high (50-100ms), maybe you should try Redis.

**Reply by jiripospisil** [reply]
Thank you, Redis works much better. I'm still getting some latency spikes but I guess that's because the machine is really slow on its own. I noticed you added support for Redis UNIX sockets in main so I will switch to that once there's a release. Cutting out the whole TCP/IP stack should help at least a little bit.

**Comment by selfhoster1312** [comment]
I started some benchmarks with garage/lmdb + juicefs/redis + qBittorrent [here](https://git.deuxfleurs.fr/Deuxfleurs/garage/issues/1021#issuecomment-12800). Unfortunately the results are really bad. I'm not sure exactly why.

Some people on torrent forums suggested that S3 is ill-suited for random writes and it would be much better to first download the torrent then "import" it to S3 for seeding.

I'm sure there's a lot of tuning that can help achieve better performance and disk usage but i don't have energy to investigate (i've never used juiceFS before). You can see my findings so far on the link i just posted. Hopefully someone who is more familiar with juiceFS tuning can find reasonable settings to make it better.

**Reply by Drallas** [reply]
My performance was awful with Backblaze, after switching to Wasabi it is very fast, torrents and nzb are blazing fast! My settings are below.

/root/.juicefs/jfsmount my-bucket /mnt/wasabi -o rw,background,max-uploads=25,prefetch=8,writeback,cache-dir=/mnt/juicefs-wasabi-cache,cache-size=102400,buffer-size=512,open-cache=true,attr-cache=10,entry-cache=10,log=/var/log/juicefs-wasabi.log,_netdev


<img width="1128" alt="Screenshot 2025-05-21 at 08 47 33" src="https://github.com/user-attachments/assets/f1a98fda-f161-4233-898f-acf66852201b" />

<img width="819" alt="Screenshot 2025-05-21 at 08 47 44" src="https://github.com/user-attachments/assets/6c99f90d-29e0-4077-babc-0e1f8d86c608" />

**Reply by selfhoster1312** [reply]
That's impressive, thanks for sharing. Can you maybe provide more information about your setup?

- How was your juiceFS volume formatted? is it using Redis / something else? what was your chosen block size?
- What is the hardware running the server, and hosting the juiceFS metadata engine? is the cache-dir on the same drive as the metadata engine?
- What torrent client are you using? With what settings?
- How many torrents are you running? How many files in total? How big are the pieces usually?

There's probably more "details" that matter but i think these information would give us a better overview. The settings i used were all documented [in this garage issue](https://git.deuxfleurs.fr/Deuxfleurs/garage/issues/1021#issuecomment-12800) but like i said i'm very new with S3 & juiceFS so i'm probably doing all the tuning wrong. (which is why i explained all the steps so a more experienced person can see where i did wrong)

**Comment by Drallas** [comment]
I'm hosting my 'Home lab' setup on a stretched docker swarm. 

1 X86 VM on Proxmox home.
3 ARM64 VPS servers. 

I'm running docker volumes and App / Media data directly from Wasabi storage, using VPS datacenter for speedy downloads. Torrents are slower, but that's more the nature of it (low reputation, not enough seeds etc etc), but NZB is quick! 

<img width="725" alt="Screenshot 2025-05-21 at 13 20 21" src="https://github.com/user-attachments/assets/f9a04e66-f1b7-4d22-ab1e-835f8c3ce78d" />


### [JuiceFS](https://juicefs.com/en/product/cloud-service) 
<img width="638" alt="Screenshot 2025-05-21 at 13 13 41" src="https://github.com/user-attachments/assets/9cab47b6-dcba-4845-96ae-72be076836bb" />

### Tailscale
Tailscale is connecting the node's, access only via Tailscale SSH, firewall is closed, apps open port(s) themselves from the inside, Cloudflare tunnels for remote access.

```
sudo ufw status verbose
Status: active
Logging: on (low)
Default: deny (incoming), allow (outgoing), deny (routed)
New profiles: skip

To                         Action      From
--                         ------      ----
Anywhere on tailscale0     ALLOW IN    Anywhere                  
Anywhere (v6) on tailscale0 ALLOW IN    Anywhere (v6)             

Anywhere                   ALLOW OUT   Anywhere on tailscale0    
Anywhere (v6)              ALLOW OUT   Anywhere (v6) on tailscale0
```

### Drawing
![Untitled-2025-05-20-1855 excalidraw](https://github.com/user-attachments/assets/d0b30cd9-bd4e-41aa-bab5-c3a35796cb6a)

**Reply by selfhoster1312** [reply]
Wow that's quite a setup! I'm still interested to know what your torrent client is and what settings it's using precisely (even if you left the default settings, it may change depending on version/distro) if you can find this information.

Thanks for sharing!

**Reply by Drallas** [reply]
I'm using Deluge, default settings.
